### ISSUE TITLE
Gather Op failing at back-prop with 3-D array and potentially causing JVM crash

### ISSUE THREAD

**Issue by partarstu** [deeplearning4j/deeplearning4j#9669]
#### Issue Description
- expected behavior : Gather Op works correctly for an array with any rank
- encountered behavior: This Op fails during back-prop (the part of the graph generated by org.nd4j.linalg.api.ops.impl.shape.Gather#doDiff()) during a Permute Op for a 3-D array. Batch size was 256. Last time reproduced with gradient NDArray having the following shape info:[3,256,256,768,196608,768,1,8192,1,99]. Gather dimension was 0, gather indices - [256,256], original array - matrix of shape [30522,768]. 

#### Version Information

* Deeplearning4j version - M2
* Platform information (OS, etc) - Windows 10 / Linux (Buster 10)
* CUDA version, if used - N/A
* NVIDIA driver version, if in use - N/A

#### Additional Information

The cause of the failure : Permute is being executed having only 2 dimensions instead of 3. Based on debug, I could identify that the root cause is incorrect calculation of dimension in org/nd4j/linalg/api/ops/impl/shape/Gather.java:182. Locally I used for some time an altered back-prop implementation which works fine. Because of a new load scheme for Custom Ops in M2 my custom implementation was ignored and this issue showed up.

Moreover, running the model on Linux caused intermittently a JVM crash. Tested with 3 different backends (`linux-x86_64`, `linux-x86_64-avx2` and `linux-x86_64-onednn-avx2`). The log of the crash is attached [here](https://gist.github.com/partarstu/3a101bce2dbb4e671c3d59e45c3fc00a) . After running the model with a custom back-prop implementation for Gather Op thus avoiding the issue described above, the JVM crash never happened again.

The only manifestation of an issue was the following error message on the console only on Windows platform (on Linux there was nothing) - 
`shape::doPermuteShapeInfo function failed: rearrange indexes are incorrect. Given permute indices must be < rank and >= 0.  Rearrange at index 2 was -1134333991`

_But the execution of the graph wasn't interrupted with this error, although the workflow was already broken._ This was probably what caused JVM crash at some point of the execution.

**Comment by agibsonccc** [comment]
@partarstu do you have a standalone reproduction?

**Comment by agibsonccc** [comment]
@partarstu just thinking this might be due to the way the op is being loaded. It probably doesn't have anything to do with the way the op is implemented just the way it's being configured. I'd appreciate a minimal graph we could run on M1.1 and M2 where I could narrow this down.

**Comment by partarstu** [comment]
@agibsonccc I might be wrong, but I don't think the way the Op is loaded affects this issue. As already mentioned in another [issue](https://github.com/eclipse/deeplearning4j/issues/9670#issuecomment-1100635367), the implementation [here](https://gist.github.com/partarstu/776b0cf91fb7d68881c71c1becabe99b) allows to workaround the issue (at least for my use case). 

The snippet of the graph which should allow you to reproduce the issue:
```
var sequenceLengthSd = sd.constant(256);
var hiddenSizeSd= sd.constant(768);
var tokenEmbeddingsMatrix = sd.var("tokenEmbeddingsMatrix ", new XavierInitScheme('c', 30522, 768), FLOAT, 30522, 768);
var inputTokenVocabIndices = sd.var("inputTokenVocabIndices  ", new XavierInitScheme('c', 512, 256), INT32, 512, 256);
var batchTokenEmbeddings = sd.gather("batchTokenEmbeddings ", tokenEmbeddingsMatrix, inputTokenVocabIndices, 0); 
var batchSize = batchTokenEmbeddings.shape().get(SDIndex.point(0)).castTo("batchSize", INT32);
var flatBatchSizeSd = batchSize.mul(sequenceLengthSd);
var hiddenLayerInputShape = sd.stack("hiddenLayerInputShape", 0, flatBatchSizeSd, hiddenSizeSd);
var layerInput = batchTokenEmbeddings.reshape(hiddenLayerInputShape);
// After that layerInput goes through self-attention graph
```

During back-prop, having `batchTokenEmbeddings-grad `and `inputTokenVocabIndices`, the native logic of Gather Op calculates incorrectly the dimensions for permutation. I think the most important for this issue is that both `inputTokenVocabIndices `and `tokenEmbeddingsMatrix ` are **matrices**, but the gradient comes back **as a 3-D array** and the current logic probably doesn't take this into account.

What concerns me is that the workflow is broken, native lib gives back the error which is printed out to console (only in Windows) but the graph execution still goes on. The execution should be immediately interrupted in order to avoid misleading back-prop results.

**Comment by agibsonccc** [comment]
@partarstu could you elaborate a bit on the graph continuing? From the c++ lib we return a status and if that status isn't 0 we automatically throw an exception at the java level. That is true for every op in the library. This sounds a bit 1 off and should be fixed. Do you have a github gist I could look at of the relevant behavior?

Also running your code the xavier init scheme can't actually be int32 data type. Do you mind fixing the init there so I can see what you're intending?

Edit: I just changed it to zeros and ran a feed forward. Just so I know what calls you're using are you using calculateGraidents(..) next or fit? Could you include that call as well?

**Comment by partarstu** [comment]
@agibsonccc regarding graph continuing - layerInput goes as keys, values and queries through projection weights of self-attention, the attention weights are calculated based on keys and queries projections and are applied to the value projections. It's a standard self-attention graph which I use for almost a year. The loss in the graph is formed based on softmax classifier.

**Comment by partarstu** [comment]
Regarding the github gist for the Op return status - do you mean the gist of the error log for JVM crash or there's some other log I could generate?

**Comment by agibsonccc** [comment]
@partarstu the error where it returns not the JVM crash. In this particular issue I'm trying to just understand the nature of the gather problem itself. My first goal is to reproduce this after my last batch of fixes to see if there's anymore issues.

**Comment by partarstu** [comment]
> Also running your code the xavier init scheme can't actually be int32 data type. Do you mind fixing the init there so I can see what you're intending?

In reality it's a placeholder which contains INT32 array of token indices from the vocab. I forgot that weights can be only floating values, sorry.

`var inputTokenVocabIndices = sd.placeHolder(INPUT_VAR_NAME, INT32, -1, sequenceLength);`

**Comment by partarstu** [comment]
> Edit: I just changed it to zeros and ran a feed forward. Just so I know what calls you're using are you using calculateGraidents(..) next or fit? Could you include that call as well?

I'm using **`fit`**

**Comment by partarstu** [comment]
> @partarstu the error where it returns not the JVM crash. In this particular issue I'm trying to just understand the nature of the gather problem itself. My first goal is to reproduce this after my last batch of fixes to see if there's anymore issues.

There's no specific stack trace, only an output to console with message :  
`shape::doPermuteShapeInfo function failed: rearrange indexes are incorrect. Given permute indices must be < rank and >= 0. Rearrange at index 2 was -1134333991`

"Rearrange at index 2 was ...." has different values based on execution. That's probably because the operation itself (based on debug) has no value at index 2 - the permute for back-prop has only 2 values for 3-D array

**Comment by agibsonccc** [comment]
@partarstu ok a problem like that is what I'd look for in the gist. What that's telling me is we have some sort of permute indices (probably derived from a variable) that are corrupt or wrong being passed in. That can then lead to a corrupt jvm crash as well.

**Comment by agibsonccc** [comment]
@partarstu could you maybe just DM me a flatbuffers model and a fit call to run? Maybe if this is proprietary use the forums direct messages. My main goal is to just get to your interpretation of the world as quickly as possible to reproduce this.

Edit:, this is probably where that permute comes from in Gather.java: 

```java
  SDVariable inputGradPermute = inputGrad.permute(permuteDims);
        inputGrad = sameDiff.scatterAdd(inputGradPermute, arg(1), permuteGrad);

        //Now, invert the permutation so axis is back where it was
        inputGrad = inputGrad.permute(invertDims);

```

**Comment by partarstu** [comment]
> @partarstu ok a problem like that is what I'd look for in the gist. What that's telling me is we have some sort of permute indices (probably derived from a variable) that are corrupt or wrong being passed in. That can then lead to a corrupt jvm crash as well.

I just rerun the training iteration again - there's no error log file present when this happens, only the console output. 

The permute dimensions is exactly what I mentioned originally in the description (org/nd4j/linalg/api/ops/impl/shape/Gather.java:182). They simply miss one additional dimension. I added in the gist my custom implementation which calculates those indices a little different. The basic idea is that probably the gradients rank should be used as the base for those calculations .

**Comment by partarstu** [comment]
> @partarstu could you maybe just DM me a flatbuffers model and a fit call to run? Maybe if this is proprietary use the forums direct messages. My main goal is to just get to your interpretation of the world as quickly as possible to reproduce this.

If you still can't reproduce it, I could try to send it. It's huge though (like 320MB)

**Comment by agibsonccc** [comment]
@partarstu google drive + DM link is fine. In the mean time here's a minimal try at reproducing this. I have this:
```java
SameDiff sd = SameDiff.create();
        long sequenceLength = 256;
        var sequenceLengthSd = sd.constant(sequenceLength);
        var hiddenSizeSd= sd.constant(768);
        var tokenEmbeddingsMatrix = sd.var("tokenEmbeddingsMatrix", new XavierInitScheme('c', 30522, 768), FLOAT, 30522, 768);
        var inputTokenVocabIndices = sd.placeHolder("inputTokenVocabIndices", INT32, -1, sequenceLength);        var batchTokenEmbeddings = sd.gather("batchTokenEmbeddings", tokenEmbeddingsMatrix, inputTokenVocabIndices, 0);
        var batchSize = batchTokenEmbeddings.shape().get(SDIndex.point(0)).castTo("batchSize", INT32);
        var flatBatchSizeSd = batchSize.mul(sequenceLengthSd);
        var hiddenLayerInputShape = sd.stack("hiddenLayerInputShape", 0, flatBatchSizeSd, hiddenSizeSd);
        var layerInput = batchTokenEmbeddings.reshape(hiddenLayerInputShape);
        var labels = sd.placeHolder("labels", INT32,10);
        TrainingConfig trainingConfig = TrainingConfig.builder()
                .dataSetFeatureMapping("inputTokenVocabIndices")
                .dataSetLabelMapping("labels")
                .updater(new Adam())
                .build();
        sd.setTrainingConfig(trainingConfig);
        DataSet d = new DataSet();

        d.setFeatures(Nd4j.zeros(sequenceLength));
        d.setLabels(Nd4j.ones(10));
        sd.fit(d);
```

With this I see an invalid reshape:
```
eshape: lengths before and after reshape should match, but got 1 vs 196608
o.n.l.c.n.o.NativeOpExecutioner - Failed to calculate output shapes for op reshape. Attempted to execute with 2 inputs, 0 outputs, 0 targs, 1 iargs, 0 bargs and 0 dargs. Inputs: [(FLOAT,[],c), (LONG,[2],c)]. Input var names: [one-var, shape_of_1]. Output var names: [batchTokenEmbeddings-grad] - Please see above message (printed out from c++) for a possible cause of error.
o.n.c.t.BaseND4JTest - TestMiscOpValidation.gatherTest: 1874 ms, threadCount: (7->7), jvmTotal=171966464, jvmMax=4294967296, totalBytes=0, maxBytes=4294967296, currPhys=823607296, maxPhys=4294967296

java.lang.RuntimeException: Op reshape with name reshape_1 failed to execute. Here is the error from c++: Op validation failed

	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.calculateOutputShape(NativeOpExecutioner.java:1672)
	at org.nd4j.linalg.api.ops.DynamicCustomOp.calculateOutputShape(DynamicCustomOp.java:696)
	at
```

This probably isn't correct. I'll see if I can get a reproducible example up and running but I'd appreciate your model. I'll get back to you tomorrow with where that ends up. Appreciated.

**Comment by partarstu** [comment]
> This probably isn't correct. I'll see if I can get a reproducible example up and running but I'd appreciate your model. I'll get back to you tomorrow with where that ends up. Appreciated.

Labels have incorrect shape. They should correspond to sequence length at least. Also input and labels are expected to be in your case [batch, sequencelength], which is [1,256].

**Comment by agibsonccc** [comment]
@partarstu I figured out a minimal example. Could you confirm that gather is missing an argument when it's invoked? See here for the fix: https://github.com/eclipse/deeplearning4j/pull/9674

**Comment by partarstu** [comment]
> @partarstu I figured out a minimal example. Could you confirm that gather is missing an argument when it's invoked? See here for the fix: #9674

@agibsonccc  The following snippet reproduces perfectly the issue, tested it just a minute ago:

```
SameDiff sd = SameDiff.create();
        long sequenceLength = 256;
        int hiddenSize = 768;
        var sequenceLengthSd = sd.constant(sequenceLength);
        var hiddenSizeSd = sd.constant(hiddenSize);
        var classesAmount = 10;
        var tokenEmbeddingsMatrix = sd.var("tokenEmbeddingsMatrix", new XavierInitScheme('c', 30522, hiddenSize), FLOAT, 30522, hiddenSize);
        var classificationWeights =
                sd.var("classificationWeights", new XavierInitScheme('c', hiddenSize, classesAmount), FLOAT, hiddenSize, classesAmount);
        var inputTokenVocabIndices = sd.placeHolder("inputTokenVocabIndices", INT32, -1, sequenceLength);
        var batchTokenEmbeddings = sd.gather("batchTokenEmbeddings", tokenEmbeddingsMatrix, inputTokenVocabIndices, 0);
        var batchSize = batchTokenEmbeddings.shape().get(SDIndex.point(0)).castTo("batchSize", INT32);
        var flatBatchSizeSd = batchSize.mul(sequenceLengthSd);
        var hiddenLayerInputShape = sd.stack("hiddenLayerInputShape", 0, flatBatchSizeSd, hiddenSizeSd);
        var layerInput = batchTokenEmbeddings.reshape(hiddenLayerInputShape);
        var labels = sd.placeHolder("labels", INT32, sequenceLength);
        var logits = sd.nn().linear(layerInput, classificationWeights, sd.zero("bias", classesAmount).convertToVariable());
        sd.loss().softmaxCrossEntropy(sd.oneHot(labels, classesAmount), logits, null);

        TrainingConfig trainingConfig = TrainingConfig.builder()
                .dataSetFeatureMapping("inputTokenVocabIndices")
                .dataSetLabelMapping("labels")
                .updater(new Adam())
                .build();
        sd.setTrainingConfig(trainingConfig);
        DataSet d = new DataSet();
        d.setFeatures(Nd4j.zeros(1, sequenceLength));
        d.setLabels(Nd4j.ones(sequenceLength));
        sd.fit(d);
```

Can you reproduce it as well?

FYI I use JDK 18, but I don't think it matters for this case.

**Comment by partarstu** [comment]
@agibsonccc Just took a look at your commit, I think you fixed another issue - I also encountered it once and I use a workaround in my code - direct conversion from int[] into NDArray. This issue causes a different error - a really thrown error which breaks the execution, not like the one to which this ticket is dedicated.

**Comment by agibsonccc** [comment]
@partarstu JFYI I ended up needing to do 2 things: 1. rewrite the gather derivative 2. add SDVariable based indexing to SDIndex to allow dynamic. This will take me another day or so to test and get right but your test case was super useful. Thanks!

**Comment by partarstu** [comment]
@agibsonccc Perfect. Glad I could help.

**Comment by agibsonccc** [comment]
@partarstu sorry this took a bit longer than I would have liked. It involved a bit more work than anticipated but surfaced a few new features for the framework which is nice. See: https://github.com/eclipse/deeplearning4j/pull/9674

**Comment by partarstu** [comment]
@agibsonccc no problem - I wasn't blocked anyway. As I can see the back-prop for Gather Op is now completely re-written. And the commit contains really quite a lot of new stuff. Cool, thanks!
